### ISSUE TITLE
UI for changing "temperature curve"

### ISSUE THREAD

**Issue by dirkschreib** [dirkschreib/SilentStudio#4]
SilentStudio allows this with command line options. This would be easy in SilentStudioMenu as well but not very suitable for a menu interface.

**Comment by dirkschreib** [comment]
Thanks for the feedback. For the fan curves I already tried something and that didn‘t worked as expected. I have another idea and it could be helpful that I now have two separate top level menu entries. 
In the meantime you just have to change one line in the code with the targetrpm values.

**Comment by MiroIsNot1337** [comment]
I have evidently done something wrong.

<img width="567" alt="Screenshot 18 01 2023 um 23 33 14 PM" src="https://user-images.githubusercontent.com/122529727/213310022-04747bda-2ddf-48a9-b854-3a5c09ce7f8c.png">

**Comment by MiroIsNot1337** [comment]
Thank you, that worked perfectly. I think it takes a few seconds until the fan gets turned on by Mac OS and then only reaching one of the threshholds of the temperature curves re-activates fan control via SilentStudioMenu. Could one of these options be the solutions?
1. Add a delay for the wake-up detection or rather until it tries to change the fan speed. Perhaps waiting 5 seconds would enable the OS to do its thing and then SilentStudioMenu can take over?
2. Add true temperature ranges instead of threshholds. So: 0-56=0, 57-64=900, 65=AUTO. I think in code this might mean every degree has to be added unless there is a more elegant solution:  0.0:"0", 1.0:"0", 2.0:"0", ... 56.0:"0", 57.0:"900", 58.0:900" ... 64.0:"900", 65.0:"AUTO"

**Comment by dirkschreib** [comment]
Thanks for the suggestion 1. This should be possible. 
I don‘t want to do point 2. There should be a hysteresis, i.e. with fixed ranges the fan speed would change every time 56° fan off, 57° fan on, 56° fan off, …

**Comment by MiroIsNot1337** [comment]
You are right, that does happen. It's probably not very good for the hardware if the fan unnecessarily turns on and off every few seconds (worst case). I think in most cases it this wouldn't happen though.

<img width="351" alt="Screenshot 18 01 2023 um 23 57 52 PM" src="https://user-images.githubusercontent.com/122529727/213313608-4bd72c39-86e7-4324-87f1-ac8bb7343582.png">

**Comment by dirkschreib** [comment]
I *assume* this is related to the 900 rpm value. Anything higher than 1300 rpm and the Studio will cool down pretty quickly.

**Comment by dirkschreib** [comment]
I had a little bit of weekend fun and included a small chart in the "other" menu:
<img width="336" alt="Bildschirm­foto 2023-01-22 um 13 39 20" src="https://user-images.githubusercontent.com/39700157/213916404-40bc1809-267f-4a51-b44f-c4c1eb282a74.png">
I still don't recommend this temperature curve with a lot of small steps. It was just an example.
Will be in the repo soon. (BTW: Still working on the UI to change the target rpm values)

**Comment by MiroIsNot1337** [comment]
The chart looks nice (and helpful)! I look forward to testing the next version.

**Comment by dirkschreib** [comment]
Just commited my latest changes. Please consider this a beta version to gather feedback.
It shows two charts for temperature and fan speed. And a list to show and edit the temperature curve and one more chart (as above) to visualize the temperature curve.

**Comment by dirkschreib** [comment]
This is how it looks with my standard config:
<img width="334" alt="Bildschirm­foto 2023-01-26 um 15 04 27" src="https://user-images.githubusercontent.com/39700157/214855832-16b22c29-f9f6-47b1-833b-9881597c98e2.png">
BTW: It is possible to remove temperature/rpm points with a left swipe and delete. If you quit the app from the menu, the new configuration will be saved.

**Comment by dirkschreib** [comment]
Just tried it with my Magic Mouse (I normally use a trackpad) and it works as well (swipe to the left, dragging doesn't work).
There is so much room for improvement once you start playing around with charts...
If you want to change the Y-axis, you can change this line: "chartYScale(domain: 40...70)" (2x)

**Comment by dirkschreib** [comment]
And thanks for the feedback. Highly appreciated!

**Comment by MiroIsNot1337** [comment]
In that case it is probably user error on my side. Where (and how) exactly do I have to swipe?

<img width="341" alt="Screenshot 26 01 2023 um 17 33 23 PM" src="https://user-images.githubusercontent.com/122529727/214895265-26ba0e1e-086f-48ce-b35f-ee30618fb72b.png">

I have expanded the temperature chart to 30-70 without problems. 20-70 creates weird display problems though that are not fixed by simply increasing minHeight:150 to something like 200. 30°C as the starting point is good enough for me. Perhaps you can clarify that it is °C for our American friends. It seems to be the default for most temperature apps, but it can't hurt.

<img width="316" alt="Screenshot 26 01 2023 um 17 43 20 PM" src="https://user-images.githubusercontent.com/122529727/214896315-c40aedae-9881-45d3-ba0f-dfc3edc13c54.png">

**Comment by MiroIsNot1337** [comment]
> And thanks for the feedback. Highly appreciated!

You are very welcome. I am so happy that you have solved the fan problem for me. I don't understand why the established (paid!) apps have no 0 rpm option or at least a wider range for the Mac Studio. It's very rewarding that you are willing and able to add functions for me and others and in such a short time, too. I am also learning how to make small changes myself with your help, which I much appreciate.

**Comment by MiroIsNot1337** [comment]
Never mind the question about the number in the lowest graph. The point was just outside of the visible area because the idle temp was too low.

<img width="317" alt="Screenshot 26 01 2023 um 18 01 07 PM" src="https://user-images.githubusercontent.com/122529727/214900267-498a0997-4885-4591-aefd-4585a526e062.png">

**Comment by dirkschreib** [comment]
Just separated the temperature sensors in the chart to analyze other sensors:
<img width="329" alt="Bildschirm­foto 2023-01-27 um 14 45 27" src="https://user-images.githubusercontent.com/39700157/215101769-b5ab4d68-75f8-4f66-9592-c019612b849c.png">

**Comment by MiroIsNot1337** [comment]
Busy weekend, but I was able to try out the newest code and everything works. I was surprised to see that I didn't have to add the rpm points again. Thank you for also adding 30°C to the temperature chart.

**Comment by dirkschreib** [comment]
All changes are stored in @AppStorage. That is a .plist file in ~/Library/Preference/SilentStudioMenu.plist

**Comment by MiroIsNot1337** [comment]
Ah, got it. That probably means I have some junk files from apps I have manually deleted in folders like that. I'll check. Thanks!

**Comment by MiroIsNot1337** [comment]
I have one more cosmetic idea for the temperature graph. Would it be possible to integrate a function where the temperature is displayed when hovering the mouse over the graph? So for example, while hovering the mouse at the lowest temperature it would display 6°C or whatever it is (seems to be a sensor error btw) right next to the point in the graph.

<img width="318" alt="Screenshot 03 02 2023 um 21 43 21 PM" src="https://user-images.githubusercontent.com/122529727/216707026-751da0a5-c376-4cbc-a9cd-ab30c9e57cbd.png">

Apart from that, I don't have any immediate ideas for improvements. The only thing I still don't get is how to delete rpm points, but that's not a big problem since I can just overwrite them with whatever I need.

**Comment by dirkschreib** [comment]
Thanks for the feedback. This is possible, just a little bit more complicated and I don‘t have time for this, at least currently. 
You have a Magic Mouse? Swipe enabled? Use one finger and swipe from right to left on the top surface of the mouse while the mouse cursor is on a temperature/rpm line.

**Comment by dirkschreib** [comment]
I have seen the low temperature readings from Tp02 as well. I *assume* it is related to sleep mode.

**Comment by MiroIsNot1337** [comment]
> Thanks for the feedback. This is possible, just a little bit more complicated and I don‘t have time for this, at least currently. You have a Magic Mouse? Swipe enabled? Use one finger and swipe from right to left on the top surface of the mouse while the mouse cursor is on a temperature/rpm line.

That explains my problem. I am using a Logitech MX Master 3S. In other words, I don't have any touchscreen or touchpad functionality as you would have with a Magic Mouse. I have to click in order to interact with anything on the screen.

As for the temperature display function, I don't think that's really required. It's more of a fancy bonus feature than anything else.

**Comment by MiroIsNot1337** [comment]
Tldr: Would it be possible to add the option 'Fan custom' (rpm can be specified in the script, perhaps default value 900) in addition to 'Fan on‘ and 'Fan off'?

There is one feature that would actually be very useful to me. Usually, I want the fan curve to be automatic because it means the fan stays mostly off. But when editing in Adobe Premiere, I wish I could manually set it to 900 rpm to avoid the fan turning on and off all the time because of temperature changes. These are very frequent. Scrubbing through the timeline requires a lot of cpu/gpu but the temperature immediately drops the moment I stop to add an effect or otherwise click around in the software.

I know I can uncheck 'Fan automatic' and set the fan to 'on,' but AUTO is faster than I need the fan to turn and too loud for my taste. 900 rpm is drowned out by the audio of whatever I am editing or even typing or clicking the mouse, so I don't mind it as much. And it is more than sufficient for the small temperature spikes I get while editing. At the moment, I have to uncheck 'Fan automatic' at just the right temperature point when the fan hits 900 rpm which is an unnecessary distraction while working.

**Comment by MiroIsNot1337** [comment]
I have become a true code monkey!

<img width="218" alt="Screenshot 10 02 2023 um 21 58 03 PM" src="https://user-images.githubusercontent.com/122529727/218196205-02ab87a6-d10f-4088-950a-bdf629ae6931.png">

        MenuBarExtra {
            Toggle(isOn: $state.automatic) {
                Text("Fan automatic")
            }.keyboardShortcut("a")
            Divider()
            Button("Fan off") {
                setFan("0")
                state.lasttemp = 0.0
            }
            .keyboardShortcut("0")
            Button("Fan 900") {
                setFan("900")
                state.lasttemp = 60.0
            }
            .keyboardShortcut("1")
            Button("Fan AUTO") {
                setFan("AUTO")
                state.lasttemp = 200.0
            }
            .keyboardShortcut("2")
            Divider()
            Button("Quit") {
                setFan("AUTO")
                NSApplication.shared.terminate(nil)
            }
            .keyboardShortcut("q")

**Comment by dirkschreib** [comment]
I removed the SwiftUI List from the list of temperature/rpm points. "delete" is now more discoverable. 
Close this issue for now.
